Allow writes to ILM shrink action target index

parkertimmins opened the pull request:
Currently, target indices produced by an ILM shrink action are not writable. This is because the source index of a shrink action must be read-only to perform the shrink. During shrink, settings are copied from source index to target index, causing the target index to also be read-only. 

This change adds a parameter to the shrink phase configuration, named `allow_write_after_shrink`. If this argument is present and true, the write block will be removed from the target index while shrinking, allowing future modifications to the target index. For BWC, this configuration parameter defaults to `false`, thus keeping the target index read-only.

parkertimmins committed:
Allow writes to ILM shrink action target index

The source index of a shrink action must be read-only.
During shrink, settings are copied from source index to target index.
Thus target index is also read-only.
This change adds shrink policy argument which makes target index writable.
Defaults to keeping read-only to preserve current behavior.

elasticsearchmachine commented:
Hi @parkertimmins, I've created a changelog YAML for you.

parkertimmins committed:
Update docs/changelog/107121.yaml

parkertimmins committed:
Fix existing tests

parkertimmins committed:
Fix some more existing test

gmarouli commented on server/src/main/java/org/elasticsearch/TransportVersions.java:
```diff
@@ -163,6 +163,7 @@ static TransportVersion def(int id) {
     public static final TransportVersion CCR_STATS_API_TIMEOUT_PARAM = def(8_622_00_0);
     public static final TransportVersion ESQL_ORDINAL_BLOCK = def(8_623_00_0);
     public static final TransportVersion ML_INFERENCE_COHERE_RERANK = def(8_624_00_0);
+    public static final TransportVersion ILM_SHRINK_WRITE = def(8_625_00_0);
```
Nit: Maybe to give a bit more context `ILM_SHRINK_ENABLE_WRITE`

gmarouli commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ShrinkAction.java:
```diff
@@ -57,16 +59,22 @@ public class ShrinkAction implements LifecycleAction {
+    public ShrinkAction(
+        @Nullable Integer numberOfShards,
+        @Nullable ByteSizeValue maxPrimaryShardSize,
+        @Nullable Boolean allowWritesOnTarget
```
My preference here would be to keep this as a boolean, when we are reading from the wire and its absent we fallback to the previous behaviour. I think the default behaviour is clear enough and this way we can avoid some null checks in the code.

What do you think?

parkertimmins replied:
I think that makes sense, but there's a scenario where it might be useful to have null. If a policy is created with the `allowWritesOnTarget` absent from the json, and we set the default value of false in the action. Then if the user calls `GET /_ilm/policy/my-policy` , we have to return the policy with the default value of false, rather than with the field absent. This might be confusing for the user. On the other hand, it might also be instructive to see the defaults included when the policy is accessed :thinking: . What do you think?

gmarouli replied:
This is a very good point. Now for this I do not have a good answer apart so I will delegate to "the wisdom of our ancestors", meaning I will check what the other actions do.

So, if my analysis is correct, please double check, optional objects are indeed nullable but optional booleans are not. This are the actions I stumbled across.
- [Delete](https://www.elastic.co/guide/en/elasticsearch/reference/current/ilm-delete.html)
- [Migrate](https://www.elastic.co/guide/en/elasticsearch/reference/current/ilm-migrate.html)
- [Searchable snapshot](https://www.elastic.co/guide/en/elasticsearch/reference/current/ilm-searchable-snapshot.html)

Please check as well, and if it did they all behave like this, I think it would make sense to keep it.

parkertimmins replied:
Ahh, good thinking to look into those :+1:  You're right, those have regular non-nullable booleans, and the defaults are returned by `GET`. So that answer that question, thanks!

parkertimmins committed:
Fix a bunch of tests with new ShrinkAction argument

parkertimmins committed:
Add separate step to make index writable

Since all shrink steps are skipped if the number of requested shards equals
the number of existing shards, the setting to make writable won't be applied.
Since the read-only is also never applied, this is only a problem if the index
is already read-only. To deal with this, move the setting from the shrinkStep
to its own step which is always run, even when the other shrink action steps
are not run.

parkertimmins committed:
Add integration tests

parkertimmins commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ShrinkAction.java:
```diff
@@ -158,11 +182,12 @@ public List<Step> toSteps(Client client, String phase, Step.StepKey nextStepKey)
             preShrinkBranchingKey,
             checkNotWriteIndex,
-            nextStepKey,
+            allowWritesKey,
```
@dakrone  pointed out that in the case where shrink does not happen due to the shard count not changing, the settings update in the ShrinkStep would not be run. This is not a problem if the input index is writable, since the ReadOnlyStep is also skipped by the conditional branch. But, if the input index is already read-only (which would be a bit weird), the shrunken index will remain read-only, even if the user set `allowWritesOnTarget=true` . 
To deal with this situation, I've moved the settings update to a separate step which is always run at the end of the ShrinkAction, even when all other steps are skipped.

parkertimmins committed:
Improvements from review

- make transport name more descriptive
- make allowWrite... boolean non-nullable

parkertimmins committed:
Rename step to allow_write_after_shrink

parkertimmins committed:
Merge branch 'main' into shrink-target-writable

parkertimmins committed:
Update docs

parkertimmins committed:
Add test for AllowWriteStep

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

parkertimmins committed:
spotless apply

parkertimmins commented on docs/reference/ilm/actions/ilm-shrink.asciidoc:
```diff
@@ -4,7 +4,7 @@
 Phases allowed: hot, warm.
 
-Sets a source index to <<index-blocks-read-only,read-only>> and shrinks it into
+<<index-blocks-write,Blocks writes>> on a source index and shrinks it into
```
The phrase "Blocks writes on a source index ..." is not as clear as "Sets a source index to read-only...", but since `index.blocks.write` is what actually gets set, rather than `index.blocks.read-only`, I think the current phrasing confusing.

nielsbauman commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ShrinkAction.java:
```diff
@@ -94,6 +99,11 @@ public ShrinkAction(StreamInput in) throws IOException {
             this.numberOfShards = null;
             this.maxPrimaryShardSize = ByteSizeValue.readFrom(in);
         }
+        if (in.getTransportVersion().onOrAfter(TransportVersions.ILM_SHRINK_ENABLE_WRITE)) {
+            this.allowWriteAfterShrink = in.readBoolean();
+        } else {
+            this.allowWriteAfterShrink = false;
+        }
```
```suggestion
        this.allowWriteAfterShrink = in.getTransportVersion().onOrAfter(TransportVersions.ILM_SHRINK_ENABLE_WRITE) && in.readBoolean();
```

parkertimmins replied:
Nice, I like it :+1:

nielsbauman commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/ShrinkActionTests.java:
```diff
@@ -66,18 +66,19 @@ protected ShrinkAction createTestInstance() {
     protected ShrinkAction mutateInstance(ShrinkAction action) {
+        boolean allowWriteAfterShrink = action.getAllowWriteAfterShrink() == false;
         if (action.getNumberOfShards() != null) {
-            return new ShrinkAction(action.getNumberOfShards() + randomIntBetween(1, 2), null);
+            return new ShrinkAction(action.getNumberOfShards() + randomIntBetween(1, 2), null, allowWriteAfterShrink);
         } else {
-            return new ShrinkAction(null, ByteSizeValue.ofBytes(action.getMaxPrimaryShardSize().getBytes() + 1));
+            return new ShrinkAction(null, ByteSizeValue.ofBytes(action.getMaxPrimaryShardSize().getBytes() + 1), allowWriteAfterShrink);
         }
```
I think we usually try to only change one field, to make sure every single field is covered in `equals`, `hashCode`, etc. I see that the original code already changed both, but while we're here, we might as well fix it.

parkertimmins replied:
Ahh, yeah that makes sense. Thank for the review!

parkertimmins replied:
Looks like it was actually not changing both fields at once in the original code. The format of this method is the bit different from the more common pattern of using a switch with case for each field, as only one of the first two parameters can be non-null. I've fixed it now to update only one field at a time, though the if statement is a little wonky looking with the `randonInBetween(0, 2) == 0`, as I didn't want to use randBoolean() and spent 50% of the time toggling a single boolean field.

nielsbauman replied:
Ah sorry I must have read the test a bit too quickly. What about doing:
```java
        Integer numberOfShards = action.getNumberOfShards();
        ByteSizeValue maxPrimaryShardSize = action.getMaxPrimaryShardSize();
        boolean allowWriteAfterShrink = action.getAllowWriteAfterShrink();

        switch (randomInt(2)) {
            case 0 -> numberOfShards = randomValueOtherThan(numberOfShards, () -> randomIntBetween(1, 100));
            case 1 -> maxPrimaryShardSize = randomValueOtherThan(
                maxPrimaryShardSize,
                () -> ByteSizeValue.ofBytes(randomIntBetween(1, 100))
            );
            case 2 -> allowWriteAfterShrink = allowWriteAfterShrink == false;
        }
        return new ShrinkAction(numberOfShards, maxPrimaryShardSize, allowWriteAfterShrink);
```
That way it still follows the "common pattern" and is still correct. The only downside is that we don't have a mutation that changes which of the first two fields is null/non-null, but I don't think that's too big of an issue as the `randomInstance()` method should take care of randomizing that. What do you think?

parkertimmins replied:
Unfortunately, that could produce an invalid object. For example if numberOfShards is  originally non-null and maxPrimaryShardSize is null, then if `case 1` is selected, both values will be non-null, which is invalid. (the only reason I know this is because I wrote something similar and then broke the test!)

nielsbauman replied:
Ahh of course 🤦🏻.
Then this is my last suggestion 😉. I've seen this in several places too. It's a bit more verbose than what you have now, but, since it aligns more with the common pattern, I think it'll be easier to read.
```java
        Integer numberOfShards = action.getNumberOfShards();
        ByteSizeValue maxPrimaryShardSize = action.getMaxPrimaryShardSize();
        boolean allowWriteAfterShrink = action.getAllowWriteAfterShrink();

        switch (randomInt(2)) {
            case 0 -> {
                numberOfShards = randomValueOtherThan(numberOfShards, () -> randomIntBetween(1, 100));
                maxPrimaryShardSize = null;
            }
            case 1 -> {
                maxPrimaryShardSize = randomValueOtherThan(
                    maxPrimaryShardSize,
                    () -> ByteSizeValue.ofBytes(randomIntBetween(1, 100))
                );
                numberOfShards = null;
            }
            case 2 -> allowWriteAfterShrink = allowWriteAfterShrink == false;
        }
        return new ShrinkAction(numberOfShards, maxPrimaryShardSize, allowWriteAfterShrink);
```

Since this is less of a clear "improvement", I'll let it up to you which way you want to go :)

parkertimmins replied:
I like it :+1:  Seems worthwhile to match the common pattern.

parkertimmins committed:
Remove plural from allowWrites for consistency

nielsbauman reviewed (commented):
I left two small comments. It looks really neat overall :)

parkertimmins committed:
Review fixes

- make only one field change per call to mutateInstance
- improve deserialization of allowWriteAfterShrink

parkertimmins committed:
Fix test

- fix some test
- fix broken ShrinkAction equals/hashcode found by tests

parkertimmins committed:
spotless apply

parkertimmins committed:
Merge branch 'main' into shrink-target-writable

parkertimmins committed:
Improve tests from review

parkertimmins committed:
Merge branch 'main' into shrink-target-writable

gmarouli commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/AllowWriteStep.java:
```diff
@@ -0,0 +1,66 @@
+public class AllowWriteStep extends UpdateSettingsStep {
+    public static final String NAME = "allow-write-after-shrink";
+
+    public static final Settings CLEAR_BLOCKS_WRITE_SETTING = Settings.builder()
```
Nit: I find that `CLEAR_WRITE_BLOCK_SETTINGS` reads a bit better than using the verb also the setting**s** points out that this is not just one setting but the whole settings objects, which I think gives us the freedom to diverge a bit from the exact setting name. Just food for thought.

gmarouli commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/AllowWriteStep.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * Make shrunken index writable, if requested, by removing {@link IndexMetadata.APIBlock#WRITE } block on the index.
+ */
+public class AllowWriteStep extends UpdateSettingsStep {
```
I would like to start the discussion about why you chose to override the `UpdateSettingsStep` and not to use it directly in the `ShrinkAction`.

parkertimmins replied:
Hmm, yeah I'm not sure making a new step made the most sense. My reasoning was that I wanted the steps of `ShinkAction` to always be the same set of steps. To do this, I pushed the branch on `allowWriteAfterShrink` down into the `AllowWriteStep`, allowing the `AllowWriteStep` to be added as the last step, whether or not `allowWriteAfterShrink` was true.

The alternative would be to use `UpdateSettingsStep` and to use `allowWriteAfterShrink` to conditonally build the set of steps with or without the `UpdateSettingsStep`. Looking around, I see that [DeleteAction](https://github.com/elastic/elasticsearch/blob/0075c1fb1e82c51c747729924ce67efef6daadeb/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DeleteAction.java#L88) (and others) behave like this. Do you think this is the way to go?

parkertimmins replied:
I went ahead moved the condition into `ShrinkAction` and replaced `AllowWriteStep` with `UpdateSettingsStep`. It adds a bit of complexity to ShrinkAction, but seems worthwhile since it doesn't involve a whole new step and a bunch of code.

gmarouli commented on docs/changelog/107121.yaml:
```diff
@@ -0,0 +1,5 @@
+pr: 107121
+summary: Allow writes to ILM shrink action target index
```
What do you think about this? Since this is going to be listed in the release change log it might benefit from a bit more context.
```suggestion
summary: Add a flag to re-enable writes on the final index after an ILM shrink action.
```

gmarouli commented on docs/changelog/107121.yaml:
```diff
@@ -0,0 +1,5 @@
+summary: Allow writes to ILM shrink action target index
+area: ILM+SLM
+type: enhancement
+issues: []
```
You can also add here the issue this is addressing
```suggestion
issues: 
  - 19723
```

gmarouli reviewed (commented):
Hey @parkertimmins , the PR is in great shape. Almost ready to merge, I would like to iterate a bit on the choice to extend the `UpdateSettingsStep` instead of reusing directly, after that we are good to go :)

parkertimmins committed:
Review improvements

parkertimmins committed:
Use UpdateSettingsStep rather than a new step type

Pull conditional branch from the AllowWriteStep type up into
the shrink action. This allows removing the AllowWriteStep type
and using UpdateSettingsStep directly

parkertimmins committed:
spotless apply

parkertimmins committed:
Merge branch 'main' into shrink-target-writable

parkertimmins commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ShrinkAction.java:
```diff
@@ -278,40 +299,52 @@ public List<Step> toSteps(Client client, String phase, Step.StepKey nextStepKey)
+        ShrunkenIndexCheckStep waitOnShrinkTakeover = new ShrunkenIndexCheckStep(isShrunkIndexKey, lastOrNextStep);
+
+        List<Step> steps = new ArrayList<>(
+            List.of(
```
This looks admittedly a bit wonky. It seemed the best way way to make a mutable literal list. I played with making `steps` immutable then copying it to add the `allowWriteAfterShrinkStep`, but didn't like that much either. :man_shrugging:

gmarouli replied:
You can also use a stream of steps, use null if the step shouldn't be there and then filter out nulls, amd convert it to list. We use it in certain places like cluster settings.

parkertimmins replied:
Nice, I like it!

parkertimmins committed:
spotless apply

parkertimmins committed:
Review improvement

parkertimmins committed:
Merge branch 'main' into shrink-target-writable

gmarouli commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ShrinkAction.java:
```diff
@@ -278,8 +300,15 @@ public List<Step> toSteps(Client client, String phase, Step.StepKey nextStepKey)
-        ShrunkenIndexCheckStep waitOnShrinkTakeover = new ShrunkenIndexCheckStep(isShrunkIndexKey, nextStepKey);
-        return Arrays.asList(
+        ShrunkenIndexCheckStep waitOnShrinkTakeover = new ShrunkenIndexCheckStep(isShrunkIndexKey, lastOrNextStep);
+        UpdateSettingsStep allowWriteAfterShrinkStep = new UpdateSettingsStep(
+            allowWriteKey,
+            nextStepKey,
+            client,
+            CLEAR_WRITE_BLOCK_SETTINGS
+        );
```
Minor, here we are creating the step no matter if we are going to use it or not. Consider only creating it if the flag requires it, by doing this you can also remove the flag check later. (Btw, I know it can be hard to spot things like this during incremental changes).

```
UpdateSettingsStep allowWriteAfterShrinkStep = allowWriteAfterShrink ? new UpdateSettingsStep(
            allowWriteKey,
            nextStepKey,
            client,
            CLEAR_WRITE_BLOCK_SETTINGS
        ) : null;
```

gmarouli reviewed (approved):
LGTM, thank you for fixing this @parkertimmins and welcome to the ILM world! I left 2 minor comments but the code appears to be good to go to me :)

parkertimmins committed:
review improvements

parkertimmins committed:
Merge branch 'main' into shrink-target-writable

parkertimmins committed:
spotless apply
